Mathlib/Analysis/Calculus/TangentCone.lean

```diff
@@ -98,30 +100,35 @@ theorem tangentCone_mono (h : s ⊆ t) : tangentConeAt 𝕜 s x ⊆ tangentConeA
   rintro y ⟨c, d, ds, ctop, clim⟩
   exact ⟨c, d, mem_of_superset ds fun n hn => h hn, ctop, clim⟩
 
-end TVS
+variable [ContinuousSMul 𝕜 E]
 
-section Normed
-variable [NormedAddCommGroup E] [NormedSpace 𝕜 E]
-variable [NormedAddCommGroup F] [NormedSpace 𝕜 F]
-variable [NormedAddCommGroup G] [NormedSpace ℝ G]
-variable {x y : E} {s t : Set E}
+theorem Filter.EventuallyEq.inv_smul_smul_cancel {α G K : Type*} [GroupWithZero G] [MulAction G K]
+    {f : α → G} {g : α → K}
+    {l : Filter α} (hg : Tendsto f l (𝓟 {0}ᶜ)) :
+    (fun x ↦ (f x)⁻¹ • f x • g x) =ᶠ[l] fun x ↦ g x := by
+  filter_upwards [hg.le_comap <| preimage_mem_comap (m := f) (mem_principal_self {0}ᶜ)] with x hx
+  exact inv_smul_smul₀ hx _
 
 /-- Auxiliary lemma ensuring that, under the assumptions defining the tangent cone,
 the sequence `d` tends to 0 at infinity. -/
 theorem tangentConeAt.lim_zero {α : Type*} (l : Filter α) {c : α → 𝕜} {d : α → E}
     (hc : Tendsto (fun n => ‖c n‖) l atTop) (hd : Tendsto (fun n => c n • d n) l (𝓝 y)) :
     Tendsto d l (𝓝 0) := by
-  have A : Tendsto (fun n => ‖c n‖⁻¹) l (𝓝 0) := tendsto_inv_atTop_zero.comp hc
-  have B : Tendsto (fun n => ‖c n • d n‖) l (𝓝 ‖y‖) := (continuous_norm.tendsto _).comp hd
-  have C : Tendsto (fun n => ‖c n‖⁻¹ * ‖c n • d n‖) l (𝓝 (0 * ‖y‖)) := A.mul B
-  rw [zero_mul] at C
-  have : ∀ᶠ n in l, ‖c n‖⁻¹ * ‖c n • d n‖ = ‖d n‖ := by
-    refine (eventually_ne_of_tendsto_norm_atTop hc 0).mono fun n hn => ?_
-    rw [norm_smul, ← mul_assoc, inv_mul_cancel₀, one_mul]
-    rwa [Ne, norm_eq_zero]
-  have D : Tendsto (fun n => ‖d n‖) l (𝓝 0) := Tendsto.congr' this C
-  rw [tendsto_zero_iff_norm_tendsto_zero]
-  exact D
+  have hc'' : Tendsto (fun n => c n) l (Bornology.cobounded _) := by
+    rwa [tendsto_norm_atTop_iff_cobounded] at hc
+  have hc' : Tendsto (fun n => (c n)⁻¹) l (𝓝[≠] 0) :=
+    Filter.tendsto_inv₀_cobounded'.comp hc''
+  replace hc : Tendsto (fun n => (c n)⁻¹) l (𝓝 0) := hc'.mono_right nhdsWithin_le_nhds
+  have B := hc.smul hd
+  rw [zero_smul] at B
+  refine B.congr' ?_
+  refine .inv_smul_smul_cancel ?_
+  refine hc''.mono_right ?_
+  simp only [le_principal_iff]
+  rw [← Bornology.isCobounded_def]
+  simp
```

Comment:
@ADedecker, I was pretty clumsy provint this; can you see a shorter way?

Reply:
```suggestion
  rw [tendsto_norm_atTop_iff_cobounded] at hc
  have : ∀ᶠ n in l, (c n)⁻¹ • (c n) • d n = d n := by
    filter_upwards [hc.eventually_ne_cobounded 0] with n hn using inv_smul_smul₀ hn _
  simpa only [← tendsto_congr' this, zero_smul] using tendsto_inv₀_cobounded.comp hc |>.smul hd
```